File: ApiVirtualTienda/Entity/Detalle.cs
using System;
using System.ComponentModel.DataAnnotations;
using System.ComponentModel.DataAnnotations.Schema;

namespace Entity
{
    public class Detalle
    {
        [DatabaseGenerated(DatabaseGeneratedOption.Identity)]
        [Key]
        public string Codigo { get; set; }
        public DateTime Fecha { get; set; }
        [Column(TypeName = "decimal(17,4)")]
        public decimal Descuento { get; set; }
        [Column(TypeName = "decimal(17,4)")]
        public decimal IVA { get; set; }
        [Column(TypeName = "decimal(17,4)")]
        public decimal ValorTotal { get; set; }
        public int Cantidad { get; set; }
        [Column(TypeName = "decimal(17,4)")]
        public decimal ValorDescuento { get; set; }
        [Column(TypeName = "decimal(17,4)")]
        public decimal ValorUnitario { get; set; }
        [NotMapped]
        public Producto Producto { get; set; }
        public string ProductoId { get; set; }
        public string FacturaCodigo { get; set; }

        public Detalle()
        {
            
        }

        public decimal CalcularDescuento()
        {
            return ValorDescuento = ( ValorUnitario * ( Descuento / 100 ));
        }

        public decimal CalcularIVA()
        {
            return IVA = ( ValorUnitario * 0.19m );
        }

        public decimal CalcularTotal()
        {
            CalcularDescuento();
            CalcularIVA();
            if(Descuento == 100)
            {
                return ValorTotal = 0;
            }
            return ValorTotal = ( IVA + (ValorUnitario - ValorDescuento ));
        }
    }
}

File: FrontEndVirtualTienda/src/app/models/detalle.ts
import { Producto } from "./producto";

export class Detalle {

  fecha: Date;
  descuento: number;
  iva: number;
  valorTotal: number;
  cantidad: number;
  valorDescuento: number;
  valorUnitario: number;
  producto: Producto;

  constructor(){
    this.producto = new Producto();
    this.fecha = new Date();
    this.descuento = 0;
    this.iva = 0;
    this.valorTotal = 0;
    this.cantidad = 0;
    this.valorDescuento = 0;
    this.valorUnitario = 0;
  }

        calcularDescuento()
        {
            this.valorDescuento = ( this.valorUnitario * ( this.descuento / 100 ));
        }

        calcularIVA()
        {
            this.iva = ( this.valorUnitario * 0.19 );
        }

        calcularTotal()
        {
            this.calcularDescuento();
            this.calcularIVA();
            if(this.descuento == 100)
            {
                return this.valorTotal = 0;
            }
            return this.valorTotal = ( this.iva + (this.valorUnitario - this.valorDescuento ));
        }


}
